Fix Flow union type annotations indentation

yamafaktory opened the pull request:
This should fix #646 🤞.

yamafaktory commented:
Ok... this needs a lot more reformatting regarding the tests.

yamafaktory commented:
@vjeux + @houshuang Should be great to know if this is going in the right direction before I refactor all the tests accordingly!

Example:

```js
/* @flow */

type Tree<T> =
  | { type: "Node", value: T, left: Tree<T>, right: Tree<T> }
  | { type: "EmptyTree" }
  | {
  mode: 'moving',
  currentActivity: Activity
  }
  | { type: "Test"}
```

Turns into:

```js
/* @flow */

type Tree<T> =
  | { type: "Node", value: T, left: Tree<T>, right: Tree<T> }
  | { type: "EmptyTree" }
  | {
      mode: "moving",
      currentActivity: Activity
  }
  | { type: "Test" };
```

houshuang commented:
Definitively better, but should the closing bracket be indented as far as the opening bracket?

```
/* @flow */

type Tree<T> =
  | { type: "Node", value: T, left: Tree<T>, right: Tree<T> }
  | { type: "EmptyTree" }
  | {
      mode: "moving",
      currentActivity: Activity
    }
  | { type: "Test" };
```

?

yamafaktory commented:
@houshuang I agree with you, this would make it more readable. As it starts to involve an opinionated indentation, it should be great to have @jlongster's views on that!

jlongster commented:
I think @houshuang, we should probably line up the closing brace as well. Thanks for this! I'll comment in the PR.

jlongster commented on src/printer.js:
```diff
@@ -626,7 +628,7 @@ function genericPrintNoParens(path, options, print) {
             leftBrace,
             indent(
-              options.tabWidth,
+              options.tabWidth * (parentIsUnionTypeAnnotation ? 2 : 1),
```
I think we need to calculate it differently. This is a special case where we are trying to line something up specifically. We want to indent it according to `| {`, which takes up 3 characters. So I think this should just be 

`options.tabWidth + (parentIsUnionTypeAnnotation ? 3 : 0)`

Because it doesn't matter if the tab width is 1 or 5; you always want to bump everything over 3 spaces because that's what you're trying to line up with.

jlongster replied:
To indent the end brace, just to a similar thing below around `rightBrace`:

`parentIsUnionTypeAnnotation ? indent(3, rightBrace), rightBrace`

I could be wrong, but something like that should work.

jlongster replied:
Actually, indentations are only applied when emitting newlines so you need to include the newline:

```js
indent(parentIsUnionTypeAnnotation ? 3 : 0,
       concat([options.bracketSpacing ? line : softline, rightBrace]))
```

Haven't actually testing this code

yamafaktory replied:
Awesome! I'll try it and fix the tests too! Yesterday evening I was trying to wrap my head around that indentation thing, playing with different values passed via the corresponding flag.

yamafaktory commented:
@jlongster I've implemented it and updated the tests accordingly. The only thing I slightly changed was the indentation size for the closing right brace in order to keep it aligned.

rhengles commented on src/printer.js:
```diff
@@ -628,15 +628,17 @@ function genericPrintNoParens(path, options, print) {
             leftBrace,
             indent(
-              options.tabWidth * (parentIsUnionTypeAnnotation ? 2 : 1),
+              options.tabWidth + (parentIsUnionTypeAnnotation ? 3 : 0),
```
@jlongster @yamafaktory

I think this line should be indented 2 spaces:

`options.tabWidth + (parentIsUnionTypeAnnotation ? 2 : 0)`

Because when indenting a normal object, the opening brace is not counted.

```javascript
{
  key: "value",
}
```

So, from the characters `| {` only the pipe and the space should be counted.

yamafaktory replied:
@rhengles I agree with you. This is how I tried to make it work in the first place with a different calculation. What's your opinion on this @jlongster?

jlongster replied:
@yamafaktory That's a good point, yep I think we should do what @rhengles said (sorry for the slow reply)

yamafaktory commented:
@jlongster this should be good to go now 👍

vjeux commented:
Thank you!
